File: internal/model/alert.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type AlertRule struct {
	ID         uint           `json:"id" gorm:"primaryKey"`
	Name       string         `json:"name" gorm:"size:128;not null"`
	ClusterID  uint           `json:"cluster_id" gorm:"index;not null"`
	Cluster    Cluster        `json:"cluster" gorm:"foreignKey:ClusterID"`
	Namespace  string         `json:"namespace" gorm:"size:64"`
	Resource   string         `json:"resource" gorm:"size:64"` // node, pod, deployment
	Metric     string         `json:"metric" gorm:"size:64;not null"` // cpu, memory, disk, restarts
	Condition  string         `json:"condition" gorm:"size:20;not null"` // >, <, >=, <=, ==, !=
	Threshold  float64        `json:"threshold" gorm:"not null"`
	Duration   string         `json:"duration" gorm:"size:20"` // 5m, 10m, 1h
	Channels   string         `json:"channels" gorm:"type:text"` // JSON array of channel IDs
	Enabled    bool           `json:"enabled" gorm:"default:true"`
	LastAlert  *time.Time     `json:"last_alert"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`
}

func (AlertRule) TableName() string {
	return "alert_rules"
}

type AlertHistory struct {
	ID         uint       `json:"id" gorm:"primaryKey"`
	RuleID     uint       `json:"rule_id" gorm:"index;not null"`
	Rule       AlertRule  `json:"rule" gorm:"foreignKey:RuleID"`
	ClusterID  uint       `json:"cluster_id" gorm:"index"`
	Cluster    Cluster    `json:"cluster" gorm:"foreignKey:ClusterID"`
	Namespace  string     `json:"namespace" gorm:"size:64"`
	Resource   string     `json:"resource" gorm:"size:128"`
	Message    string     `json:"message" gorm:"type:text"`
	Value      float64    `json:"value"`
	Status     string     `json:"status" gorm:"size:20;default:'firing'"` // firing, resolved
	TriggeredAt time.Time `json:"triggered_at" gorm:"index"`
	ResolvedAt *time.Time `json:"resolved_at"`
	Notified   bool       `json:"notified"`
	NotifyAt   *time.Time `json:"notify_at"`
}

func (AlertHistory) TableName() string {
	return "alert_history"
}

type NotificationChannel struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	Name      string         `json:"name" gorm:"size:64;not null"`
	Type      string         `json:"type" gorm:"size:20;not null"` // email, webhook, dingtalk, wechat
	Config    string         `json:"config" gorm:"type:text"` // JSON config
	Enabled   bool           `json:"enabled" gorm:"default:true"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

func (NotificationChannel) TableName() string {
	return "notification_channels"
}

File: frontend/src/api/alert.ts
import { get, post, put, del } from './request'

export interface AlertRule {
  id: number
  name: string
  cluster_id: number
  cluster?: { name: string; display_name?: string }
  namespace: string
  resource: string
  metric: string
  condition: string
  threshold: number
  duration: string
  channels: string
  enabled: boolean
  last_alert: string | null
  last_eval_at?: string | null
  last_eval_error?: string
  created_at: string
}

export interface AlertHistory {
  id: number
  rule_id: number
  rule?: { name: string }
  cluster_id: number
  cluster?: { name: string; display_name?: string }
  namespace: string
  resource: string
  message: string
  value: number
  status: string
  triggered_at: string
  resolved_at: string | null
  notified?: boolean
}

export interface NotificationChannel {
  id: number
  name: string
  type: string
  config: string
  enabled: boolean
  created_at: string
}

export const parseChannelIDs = (raw?: string): number[] => {
  if (!raw) return []
  try {
    const parsed = JSON.parse(raw)
    return Array.isArray(parsed) ? parsed.map(Number).filter((n) => !Number.isNaN(n)) : []
  } catch {
    return []
  }
}

// Alert Rules
export const getAlertRules = (page = 1, size = 10, clusterId?: number) => {
  const params: Record<string, number> = { page, size }
  if (clusterId) params.cluster_id = clusterId
  return get<{ code: number; data: AlertRule[]; total: number }>('/alerts/rules', { params })
}

export const createAlertRule = (data: {
  name: string
  cluster_id: number
  namespace?: string
  resource?: string
  metric: string
  condition: string
  threshold: number
  duration?: string
  channels?: number[]
  enabled?: boolean
}) => {
  return post('/alerts/rules', data)
}

export const updateAlertRule = (id: number, data: {
  name?: string
  namespace?: string
  resource?: string
  metric?: string
  condition?: string
  threshold?: number
  duration?: string
  channels?: number[]
  enabled?: boolean
}) => {
  return put(`/alerts/rules/${id}`, data)
}

export const deleteAlertRule = (id: number) => {
  return del(`/alerts/rules/${id}`)
}

// Alert History
export const getAlertHistory = (page = 1, size = 20, status?: string) => {
  const params: Record<string, number | string> = { page, size }
  if (status) params.status = status
  return get<{ code: number; data: AlertHistory[]; total: number }>('/alerts/history', { params })
}

// Notification Channels
export const getNotificationChannels = () => {
  return get<{ code: number; data: NotificationChannel[] }>('/alerts/channels')
}

export const createNotificationChannel = (data: {
  name: string
  type: string
  config: string
  enabled?: boolean
}) => {
  return post('/alerts/channels', data)
}

export const updateNotificationChannel = (id: number, data: {
  name?: string
  type?: string
  config?: string
  enabled?: boolean
}) => {
  return put(`/alerts/channels/${id}`, data)
}

export const deleteNotificationChannel = (id: number) => {
  return del(`/alerts/channels/${id}`)
}
